### ISSUE TITLE
onBlur and onClick cancelling each other

### ISSUE THREAD

**Issue by AdamModus** [szhsin/react-menu#838]
React/React-dom version: ^16.13.1
React-Menu version: 3.4.1

## Describe the bug
I am in a situation where I need to use my own button component. Due to that, I used the ControlledMenu and then the useMenuState to help me control the state of the menu.
I am also rendering my menu in a portal.

However, I have noticed that if I use the onClose to leverage the closing of the menu, this will conflict with the Menu Button. It causes the Menu to stay open.

As a user, if the Menu is open and I clicked on the Menu Button, I would expect the Menu to close.

## To Reproduce
1. I have a reproducible code sandbox here: https://codesandbox.io/s/react-menu-starter-forked-x1vmvi?file=/src/App.js
2. Check the console logs to see the order of events triggered.
3. Click the Menu Button to open the Menu
4. Click the Menu Button to close the Menu
5. Notice Menu Button remains open

## Expected behaviour
Clicking outside of the Menu, including the menu button, closes the menu. In these situations, I believe it is best to let the dev who is implementing the menu leverage the state.

## Code snippets/CodeSandbox examples/Screenshots
https://codesandbox.io/s/react-menu-starter-forked-x1vmvi?file=/src/App.js

## Results of my own investigation
I looked into the codebase to try to understand what is happening and maybe help you to figure out a solution. I see that in the file MenuContainer, the onBlur is being triggered due to clicking on the Menu Button. This makes complete sense.

However, I see that the check being made there is for noticing if the blurred element (the Menu itself) does not contain the related element (in this case, the Menu Button). Due to the fact that the Menu is rendered in a portal, it will **never** contain the Menu Button, thus always triggering the onClose being called.

I see 3 potential solutions:

1. I believe the correct behaviour should be: when a dev provides their custom button (via AnchorRef), to see if the related element (if it is that AncherRed element), to not call the onClose.
2. Another option would be to extend the API of the ControlledMenu to allow a dev to give the ControlledMenu the ability to provide their own buttons to it.
3. Potentially the easiest, smallest code change for you. When onBlur is triggered, call onClose with the event information, so that the dev can make the decision by themselves (do the check I mentioned in point 1).

Example of the solution on point 3:
```
  const onBlur = (e) => {
    if (isOpen && !e.currentTarget.contains(e.relatedTarget || document.activeElement)) {
      e.reason = CloseReason.BLUR;
      safeCall(onClose, e);

      // If a user clicks on the menu button when a menu is open, we need to close the menu.
      // However, a blur event will be fired prior to the click event on menu button,
      // which makes the menu first close and then open again.
      // If this happen, e.relatedTarget is incorrectly set to null instead of the button in Safari and Firefox,
      // and makes it difficult to determine whether onBlur is fired because of clicking on menu button.
      // This is a workaround approach which sets a flag to skip a following click event.
      if (skipOpen) {
        skipOpen.current = true;
        setTimeout(() => (skipOpen.current = false), 300);
      }
    }
  };
```

I am happy to provide further info if needed. Or even set a call or anything like that! I want to contribute as much as possible.

react-menu overall is a great project and I really enjoy it!

**Comment by szhsin** [comment]
Hi @AdamModus , thank you for the elaborate report and sharing your thoughts. 

The issue is due to `onBlur` event has inconsistent behaviours across browsers. The `Menu` component applies a [workaround](https://github.com/szhsin/react-menu/blob/76a84da01875628055cce1102b94c2209d893dfa/src/components/MenuContainer.js#L37) to overcome it. `ControlledMenu` doesn't contain the workaround but it can be implemented in user code in a similar way.

But before we look into that, you can first try to use the `Menu` directly as it works with button component as well [sandbox example](https://codesandbox.io/s/react-menu-material-ui-example-wvzpc?file=/src/App.js).

If your own button component doesn't accept ref or click event, there is also a way to [get around](https://github.com/szhsin/react-menu/issues/693#issuecomment-1290304080) it.

Let's get back to the `onBlur` issue when using `ControlledMenu`:
- Portal doesn't play a factor here because the button always renders outside menu container regardless whether the `portal` props is set. In addition, React portal doesn't change [event bubbling](https://reactjs.org/docs/portals.html#event-bubbling-through-portals).
- You can use a workaround (**_not recommended_**) that is the same as `Menu` currently, see [sandbox example](https://codesandbox.io/s/react-menu-starter-forked-wywlss?file=/src/App.js). However, the timer is not ideal and there is another workaround a lot better (next point).
- It just happens that I'm working on a [new `useClick` hook](https://github.com/szhsin/react-menu/blob/5f1d07e572be563c1fae640cba9f06a6a9cc5eec/src/hooks/useClick.js) that uses another method to workaround it, by using two events (`onMouseDown` and `onClick`). The hook can work with `ControlledMenu` to fix the exact issue you are facing. Code change has already been done and I'm updating the docs and examples, so it would be released in the next version soon.

**Comment by AdamModus** [comment]
Hi @szhsin , thanks for the quick response!

Unfortunately, in my situation, I cannot use the `Menu` directly, as the developers I'm serving need explicit control (the menu will be used in many different ways). So `ControlledMenu` will have to stay.

I had looked into trying to use `skipOpen` but, because it wasn't mentioned in the documentation and I felt like there likely were better solutions, I ruled it out and thought of reaching out here first 🙂. I have built a not very smaller abstraction on top of react-menu, so using something that I did not know would stay available/be supported would make code on my side potentially brittle.

The useClick hook looks good! Thank you for sharing that! I believe that, within the abstractions on my side, using that hook would work.
If there is anything I can do to speed that up, please let me know, as it became a "hard dependency" on my side.

**Comment by AdamModus** [comment]
@szhsin, just to know, in case it is possible, is there a release date for the next version?

**Comment by szhsin** [comment]
@AdamModus, the next version is coming very soon and could be available in a few days.

**Comment by nik72619c** [comment]
@szhsin Thanks for the help! I believe we can have this fixed in the next version by mid next week?

**Comment by szhsin** [comment]
@nik72619c you can surely count on it.

**Comment by szhsin** [comment]
Released in [v3.5.0](https://github.com/szhsin/react-menu/releases/tag/v3.5.0)

**Comment by AdamModus** [comment]
@szhsin , just adjusted my code as per your latest release. It works perfectly. Thanks! 🔥
